src/btree/bt_curnext.c

```diff
@@ -430,6 +438,104 @@ __wt_btcur_iterate_setup(WT_CURSOR_BTREE *cbt)
 	}
 }
 
+#ifdef HAVE_DIAGNOSTIC
+/*
+ * __cursor_key_order_check_col --
+ *	Check key ordering for column-store cursor movements.
+ */
+static int
+__cursor_key_order_check_col(
+    WT_SESSION_IMPL *session, WT_CURSOR_BTREE *cbt, bool next)
+{
+	int cmp;
+
+	cmp = 0;			/* -Werror=maybe-uninitialized */
+
+	if (cbt->lastrecno != WT_RECNO_OOB) {
+		if (cbt->lastrecno < cbt->recno)
+			cmp = -1;
+		if (cbt->lastrecno > cbt->recno)
+			cmp = 1;
+	}
+
+	if (cbt->lastrecno == WT_RECNO_OOB ||
+	    (next && cmp < 0) || (!next && cmp > 0)) {
+		cbt->lastrecno = cbt->recno;
+		return (0);
+	}
+
+	WT_PANIC_RET(session, EINVAL,
+	    "WT_CURSOR.%s out-of-order returns: returned key %" PRIu64 " then "
+	    "key %" PRIu64,
+	    next ? "next" : "prev", cbt->lastrecno, cbt->recno);
+}
+
+/*
+ * __cursor_key_order_check_row --
+ *	Check key ordering for row-store cursor movements.
+ */
+static int
+__cursor_key_order_check_row(
+    WT_SESSION_IMPL *session, WT_CURSOR_BTREE *cbt, bool next)
+{
+	WT_BTREE *btree;
+	WT_ITEM *key;
+	WT_DECL_RET;
+	WT_DECL_ITEM(a);
+	WT_DECL_ITEM(b);
+	int cmp;
+
+	btree = S2BT(session);
+	key = &cbt->iface.key;
+	cmp = 0;			/* -Werror=maybe-uninitialized */
+
+	if (cbt->lastkey->size != 0)
+		WT_RET(__wt_compare(
+		    session, btree->collator, cbt->lastkey, key, &cmp));
+
+	if (cbt->lastkey->size == 0 || (next && cmp < 0) || (!next && cmp > 0))
+		return (__wt_buf_set(session, cbt->lastkey,
+		    cbt->iface.key.data, cbt->iface.key.size));
+
+	WT_ERR(__wt_scr_alloc(session, 512, &a));
+	WT_ERR(__wt_buf_set_printable(
+	    session, a, cbt->lastkey->data, cbt->lastkey->size));
+
+	WT_ERR(__wt_scr_alloc(session, 512, &b));
+	WT_ERR(__wt_buf_set_printable(session, b, key->data, key->size));
+
+	WT_PANIC_ERR(session, EINVAL,
+	    "WT_CURSOR.%s out-of-order returns: returned key %.*s then "
```

Comment:
@sueloverso and I agreed this information is probably available from the key, so we'll leave the current messages unchanged.